File: pubnub-gson/pubnub-gson-docs/src/main/java/com/pubnub/docs/publishAndSubscribe/eventListener/EventListenersOthers.java
package com.pubnub.docs.publishAndSubscribe.eventListener;

import com.pubnub.api.PubNubException;
import com.pubnub.api.java.PubNub;
import com.pubnub.api.java.models.consumer.objects_api.channel.PNChannelMetadataResult;
import com.pubnub.api.java.models.consumer.objects_api.membership.PNMembershipResult;
import com.pubnub.api.java.models.consumer.objects_api.uuid.PNUUIDMetadataResult;
import com.pubnub.api.java.v2.callbacks.EventListener;
import com.pubnub.api.java.v2.subscriptions.Subscription;
import com.pubnub.api.java.v2.subscriptions.SubscriptionSet;
import com.pubnub.api.models.consumer.pubsub.PNMessageResult;
import com.pubnub.api.models.consumer.pubsub.PNPresenceEventResult;
import com.pubnub.api.models.consumer.pubsub.PNSignalResult;
import com.pubnub.api.models.consumer.pubsub.files.PNFileEventResult;
import com.pubnub.api.models.consumer.pubsub.message_actions.PNMessageActionResult;
import com.pubnub.api.v2.subscriptions.SubscriptionOptions;
import com.pubnub.docs.SnippetBase;
import org.jetbrains.annotations.NotNull;

import java.util.HashSet;
import java.util.Set;

public class EventListenersOthers extends SnippetBase {
    private void addEventListeners() throws PubNubException {
        // https://www.pubnub.com/docs/sdks/java/api-reference/publish-and-subscribe#basic-usage-5

        PubNub pubNub = createPubNub();


        // snippet.addEventListeners
        // Create a subscription to a specific channel
        Subscription subscription = pubNub.channel("my_channel").subscription(SubscriptionOptions.receivePresenceEvents());

        subscription.addListener(new EventListener() {
            @Override
            public void message(@NotNull PubNub pubnub, @NotNull PNMessageResult result) {
                // Log or process message
                System.out.println("Message: " + result.getMessage());
            }

            @Override
            public void presence(@NotNull PubNub pubnub, @NotNull PNPresenceEventResult result) {
                // Handle presence updates
                // requires a subscription with presence
                System.out.println("Presence userId: " + result.getUuid() + ", Event: " + result.getEvent());
            }

            @Override
            public void signal(@NotNull PubNub pubnub, @NotNull PNSignalResult result) {
                // Handle signals
                System.out.println("Signal: " + result.getMessage());
            }

            @Override
            public void messageAction(@NotNull PubNub pubNub, @NotNull PNMessageActionResult result) {
                // Handle message reactions
                System.out.println("Message Reaction: " + result.getData());
            }

            @Override
            public void file(@NotNull PubNub pubnub, @NotNull PNFileEventResult result) {
                // Handle file events
                System.out.println("File: " + result.getFile().getName());
            }

            @Override
            public void uuid(@NotNull PubNub pubnub, @NotNull PNUUIDMetadataResult pnUUIDMetadataResult) {
                // Handle uuid metadata events
                System.out.println("UUID: " + pnUUIDMetadataResult.getData().getName());
            }

            @Override
            public void channel(@NotNull PubNub pubnub, @NotNull PNChannelMetadataResult pnChannelMetadataResult) {
                // Handle channel metadata events
                System.out.println("Channel: " + pnChannelMetadataResult.getData().getName());
            }

            @Override
            public void membership(@NotNull PubNub pubnub, @NotNull PNMembershipResult pnMembershipResult) {
                // Handle membership metadata events
                System.out.println("Channel: " + pnMembershipResult.getData().getChannel().getName() + " UUID: "
                        + pnMembershipResult.getData().getUuid());
            }
        });

        // Activate the subscription to start receiving events
        subscription.subscribe();

        // Print a status when successfully subscribed
        System.out.println("Subscribed to channel 'my_channel'");


        Subscription subscription02 = pubNub.channel("my_channel02").subscription();
        Set<Subscription> subscriptionsSet = new HashSet<>();
        subscriptionsSet.add(subscription);
        subscriptionsSet.add(subscription02);

        SubscriptionSet subscriptionSet = pubNub.subscriptionSetOf(subscriptionsSet);

        // add global listener that listen on all subscribed channels
        pubNub.addListener(new EventListener() {
            @Override
            public void message(@NotNull PubNub pubnub, @NotNull PNMessageResult message) {
                System.out.println("Message: " + message.getMessage());
            }
        });

        // Activate the subscriptionSet to start receiving events
        subscriptionSet.subscribe();
        // snippet.end
    }
}

File: pubnub-kotlin/pubnub-kotlin-docs/src/main/kotlin/com/pubnub/docs/publishAndSubscribe/eventListeners/EventListenersOthers.kt
package com.pubnub.docs.publishAndSubscribe.eventListeners

import com.pubnub.api.PubNub
import com.pubnub.api.models.consumer.PNStatus
import com.pubnub.api.models.consumer.pubsub.PNMessageResult
import com.pubnub.api.models.consumer.pubsub.PNPresenceEventResult
import com.pubnub.api.models.consumer.pubsub.PNSignalResult
import com.pubnub.api.models.consumer.pubsub.files.PNFileEventResult
import com.pubnub.api.models.consumer.pubsub.message_actions.PNMessageActionResult
import com.pubnub.api.models.consumer.pubsub.objects.PNObjectEventResult
import com.pubnub.api.v2.callbacks.EventListener
import com.pubnub.api.v2.callbacks.StatusListener
import com.pubnub.docs.SnippetBase

class EventListenersOthers : SnippetBase() {
    fun addListenersBasic() {
        // https://www.pubnub.com/docs/sdks/kotlin/api-reference/publish-and-subscribe#basic-usage-5

        val pubnub = createPubNub()

        // snippet.addListenersBasic
        // Create a subscription to a specific channel
        val subscription = pubnub.channel("my_channel").subscription()

        // Add a listener to the subscription for handling various event types
        subscription.addListener(object : EventListener {
            override fun message(pubnub: PubNub, message: PNMessageResult) {
                // Log or process message
                println("Message: ${message.message}")
            }

            override fun signal(pubnub: PubNub, signal: PNSignalResult) {
                // Handle signals
                println("Signal: ${signal.message}")
            }

            override fun messageAction(pubnub: PubNub, messageAction: PNMessageActionResult) {
                // Handle message reactions
                println("Message Reaction: ${messageAction.data}")
            }

            override fun file(pubnub: PubNub, file: PNFileEventResult) {
                // Handle file events
                println("File: ${file.file.name}")
            }

            override fun objects(pubnub: PubNub, obj: PNObjectEventResult) {
                // Handle metadata updates
                println("App Context: ${obj.extractedMessage.event}")
            }

            override fun presence(pubnub: PubNub, presence: PNPresenceEventResult) {
                // Handle presence updates
                // requires a subscription with presence
                println("Presence: ${presence.uuid} - ${presence.event}")
            }
        })

        // Activate the subscription to start receiving events
        subscription.subscribe()

        // Print a status when successfully subscribed
        println("Subscribed to channel 'my_channel'")

        // Create subscription set
        val subscriptionSet = pubnub.subscriptionSetOf(
            // Specify channels with default options
            channels = setOf("my_channel", "other_channel"),
        )

        // Add listener to the subscriptionSet
        subscriptionSet.addListener(object : EventListener {
            override fun message(pubnub: PubNub, message: PNMessageResult) {
                // Log or process message
                println("Message: ${message.message}")
            }
        })

        // Activate the subscriptionSet to start receiving events
        subscriptionSet.subscribe()

        // snippet.end
    }

    private fun createListenerForOneTypeEvent() {
        // https://www.pubnub.com/docs/sdks/kotlin/api-reference/publish-and-subscribe#basic-usage-6

        val pubnub = createPubNub()
        val subscription = pubnub.channel("my_channel").subscription()

        // snippet.createListenerForOneTypeEvent
        subscription.onMessage = { message ->
            // Handle message
        }

        subscription.onSignal = { signal ->
            // Handle signal
        }

        subscription.onMessageAction = { messageAction ->
            // Handle message reaction
        }

        subscription.onFile = { file ->
            // Handle file event
        }

        subscription.onObjects = { obj ->
            // Handle metadata updates
        }

        subscription.onPresence = { presence ->
            // Handle presence updates
        }

        val onMessage: (PNMessageResult) -> Unit = { /* Handle message */ }
        val onSignal: (PNSignalResult) -> Unit = { /* Handle signal */ }
        val onMessageAction: (PNMessageActionResult) -> Unit = { /* Handle message reaction */ }
        val onFile: (PNFileEventResult) -> Unit = { /* Handle file event */ }
        val onObjects: (PNObjectEventResult) -> Unit = { /* Handle metadata updates */ }
        val onPresence: (PNPresenceEventResult) -> Unit = { /* Handle presence updates */ }

        subscription.onMessage = onMessage
        subscription.onSignal = onSignal
        subscription.onMessageAction = onMessageAction
        subscription.onFile = onFile
        subscription.onObjects = onObjects
        subscription.onPresence = onPresence

        // snippet.end

        // https://www.pubnub.com/docs/sdks/kotlin/api-reference/publish-and-subscribe#remove-event-listener

        // snippet.removeEventListener
        subscription.onMessage = null
        subscription.onSignal = null
        subscription.onMessageAction = null
        subscription.onFile = null
        subscription.onObjects = null
        subscription.onPresence = null

        // snippet.end
    }

    private fun addConnectionStatusListenerMethod() {
        // https://www.pubnub.com/docs/sdks/kotlin/api-reference/publish-and-subscribe#methods-6

        val pubnub = createPubNub()

        // snippet.addConnectionStatusListenerMethod
        pubnub.addListener(object : StatusListener {
            override fun status(pubnub: PubNub, status: PNStatus) {
                // Handle connection status updates
                println("Connection Status: ${status.category}")
            }
        })

        // snippet.end
    }
}
